File: java/src/main/java/io/cucumber/query/LineageReducerDescending.java
package io.cucumber.query;

import io.cucumber.messages.types.Pickle;

import java.util.function.Supplier;

import static java.util.Objects.requireNonNull;

/**
 * Reduces the lineage of a Gherkin document element in descending order.
 *
 * @param <T> type to which the lineage is reduced.
 */
class LineageReducerDescending<T> implements LineageReducer<T> {

    private final Supplier<? extends LineageCollector<T>> reducerSupplier;

    LineageReducerDescending(Supplier<? extends LineageCollector<T>> reducerSupplier) {
        this.reducerSupplier = requireNonNull(reducerSupplier);
    }

    @Override
    public T reduce(Lineage lineage) {
        LineageCollector<T> reducer = reducerSupplier.get();
        reduceAddLineage(reducer, lineage);
        return reducer.finish();
    }

    @Override
    public T reduce(Lineage lineage, Pickle pickle) {
        LineageCollector<T> reducer = reducerSupplier.get();
        reduceAddLineage(reducer, lineage);
        reducer.add(pickle);
        return reducer.finish();
    }

    private static <T> void reduceAddLineage(LineageCollector<T> reducer, Lineage lineage) {
        reducer.add(lineage.document());
        lineage.feature().ifPresent(reducer::add);
        lineage.rule().ifPresent(reducer::add);
        lineage.scenario().ifPresent(reducer::add);
        lineage.examples().ifPresent(examples -> reducer.add(examples, lineage.examplesIndex().orElse(0)));
        lineage.example().ifPresent(example -> reducer.add(example, lineage.exampleIndex().orElse(0)));
    }
}

File: dotnet/Query/Query/LineageReducerDescending.cs
using System;
using Io.Cucumber.Messages.Types;

namespace Io.Cucumber.Query
{
    // port of io.cucumber.query.LineageReducerDescending (Java)
    public class LineageReducerDescending<T> : ILineageReducer<T>
    {
        private readonly Func<ICollector<T>> _collectorSupplier;

        public LineageReducerDescending(Func<ICollector<T>> collectorSupplier)
        {
            _collectorSupplier = collectorSupplier ?? throw new ArgumentNullException(nameof(collectorSupplier));
        }

        public T Reduce(Lineage lineage)
        {
            var collector = _collectorSupplier();
            ReduceAddLineage(collector, lineage);
            return collector.Finish();
        }

        public T Reduce(Lineage lineage, Pickle pickle)
        {
            var collector = _collectorSupplier();
            ReduceAddLineage(collector, lineage);
            collector.Add(pickle);
            return collector.Finish();
        }

        private static void ReduceAddLineage(ICollector<T> collector, Lineage lineage)
        {
            if (lineage.Document != null)
                collector.Add(lineage.Document);
            if (lineage.Feature != null)
                collector.Add(lineage.Feature);
            if (lineage.Rule != null)
                collector.Add(lineage.Rule);
            if (lineage.Scenario != null)
                collector.Add(lineage.Scenario);
            if (lineage.Examples != null)
                collector.Add(lineage.Examples, lineage.ExamplesIndex ?? 0);
            if (lineage.Example != null)
                collector.Add(lineage.Example, lineage.ExampleIndex ?? 0);
        }
    }
}
